api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -336,7 +329,47 @@ public static void validateWriteSchema(Schema tableSchema, Schema writeSchema,
       for (String error : errors) {
         sb.append("\n* ").append(error);
       }
-      throw new IllegalArgumentException(sb.toString());
+      return sb.toString();
+    });
+  }
+
+  /**
+   * Validates whether the provided schema is compatible with the expected schema.
+   *
+   * @param ctx the validation context used in the error message
+   * @param expectedSchema the expected schema
+   * @param actualSchema the actual schema
+   * @param checkNullability whether to check nullability
+   * @param checkOrdering whether to check field ordering
+   */
+  public static void validateSchema(String ctx, Schema expectedSchema, Schema actualSchema,
+                                    Boolean checkNullability, Boolean checkOrdering) {
+    validateSchema(expectedSchema, actualSchema, checkNullability, checkOrdering, (errors) -> {
+      StringBuilder sb = new StringBuilder();
+      String header = String.format("Schema is not compatible with expected %s schema:", ctx);
+      sb.append(header)
+          .append(expectedSchema)
+          .append("\nactual schema:")
+          .append(actualSchema)
+          .append("\nproblems:");
+      for (String error : errors) {
+        sb.append("\n* ").append(error);
+      }
+      return sb.toString();
+    });
+  }
+
+  private static void validateSchema(Schema expectedSchema, Schema actualSchema, Boolean checkNullability,
```

Comment:
I did this refactoring to slightly reduce the code duplication. Now I look at it and I am not sure it was worth it.

Reply:
Yeah, seems like just adding the context part is all you need to do.